chia/consensus/multiprocess_validation.py

```diff
@@ -221,10 +222,18 @@ async def pre_validate_blocks_multiprocessing(
     diff_ssis: List[Tuple[uint64, uint64]] = []
     for block in blocks:
         if block.height != 0:
-            assert block_records.contains_block(block.prev_header_hash)
             if prev_b is None:
-                prev_b = block_records.block_record(block.prev_header_hash)
+                prev_b = await block_records.get_block_record_from_db(block.prev_header_hash)
+            assert prev_b is not None
 
+        # get_next_sub_slot_iters_and_difficulty() requires prev_b.prev_hash to
+        # be in the block record cache.
+        if prev_b is not None and prev_b.height != 0 and not block_records.contains_block(prev_b.prev_hash):
+            return [PreValidationResult(uint16(Err.INVALID_PREV_BLOCK_HASH.value), None, None, False)]
```

Comment:
There are some existing tests that are sensitive to always allowing falling back to the database. Specifically, if this check is removed (always populating the cache if the block record is missing), `test_full_node.py::test_long_reorg` fails:
```
tests/core/full_node/test_full_node.py:2231: in test_long_reorg
    await node.full_node.add_block(b, fork_info=fork_info)
chia/full_node/full_node.py:1687: in add_block
    pre_validation_results = await self.blockchain.pre_validate_blocks_multiprocessing(
chia/consensus/blockchain.py:818: in pre_validate_blocks_multiprocessing
    return await pre_validate_blocks_multiprocessing(
chia/consensus/multiprocess_validation.py:237: in pre_validate_blocks_multiprocessing
    sub_slot_iters, difficulty = get_next_sub_slot_iters_and_difficulty(
chia/consensus/difficulty_adjustment.py:374: in get_next_sub_slot_iters_and_difficulty
    prev_difficulty: uint64 = uint64(prev_b.weight - blocks.block_record(prev_b.prev_hash).weight)
chia/consensus/blockchain.py:856: in block_record
    return self.__block_records[header_hash]
E   KeyError: <bytes32: 55c436191c14810764e81d9d8d0bd7b229037e5213ae4019680974d74286df85>
```

Reply:
that was one of my tests, so not the best example. Here's an existing test that fails:

`test_full_sync.py::TestFullSync::test_sync_bad_peak_while_synced`:
```
tests/core/full_node/full_sync/test_full_sync.py:325: in test_sync_bad_peak_while_synced
    await full_node_2.full_node.add_block(block)
chia/full_node/full_node.py:1687: in add_block
    pre_validation_results = await self.blockchain.pre_validate_blocks_multiprocessing(
chia/consensus/blockchain.py:818: in pre_validate_blocks_multiprocessing
    return await pre_validate_blocks_multiprocessing(
chia/consensus/multiprocess_validation.py:237: in pre_validate_blocks_multiprocessing
    sub_slot_iters, difficulty = get_next_sub_slot_iters_and_difficulty(
chia/consensus/difficulty_adjustment.py:374: in get_next_sub_slot_iters_and_difficulty
    prev_difficulty: uint64 = uint64(prev_b.weight - blocks.block_record(prev_b.prev_hash).weight)
chia/consensus/blockchain.py:856: in block_record
    return self.__block_records[header_hash]
E   KeyError: <bytes32: e1a54d7dfe46d1770695bf02acf340507b40ee6d21168daea3e92319c1d488b6>
```

Reply:
> as i understand if you only fall back to DB in some cases that means that you can get a Key errors on a long enogh reorg

I added tests to cover long reorgs, and carving out this `INVALID_PREV_BLOCK_HASH` failure does not prevent them for passing.

> if that is not the case then we dont really need the fall back in the rest of the places right ?

We do, otherwise the long reorgs fail. Note that this is a very specific condition where a specific block needs to be in the cache. Once we start validating, we add the block records to the cache.

It *is* possible that an alternative approach would be to warm-up the cache when we detect any reorg, and then rely on block records being added as we validate them.